service/topology_coordinator.cc

```diff
@@ -752,6 +754,83 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+            while (true) {
+                auto tmptr = get_token_metadata_ptr();
+                sstring ks_name = *_topo_sm._topology.new_keyspace_rf_change_ks_name;
+                std::unordered_map<sstring, sstring> new_rf_per_dc = *_topo_sm._topology.new_keyspace_rf_change_rf_per_dc;
+                std::map<sstring, sstring> new_rf_per_dc_map{new_rf_per_dc.begin(), new_rf_per_dc.end()};
+                sstring req_uuid = *_topo_sm._topology.global_request_id;
+                std::vector<canonical_mutation> updates;
+                sstring error;
+                auto& ks = _db.find_keyspace(ks_name);
+                size_t unimportant_init_tablet_count = 2; // must be a power of 2
+                locator::tablet_map new_tablet_map{unimportant_init_tablet_count};
+
+                for (const auto& table : ks.metadata()->tables()) {
+                    try {
+                        locator::tablet_map old_tablets = tmptr->tablets().get_tablet_map(table->id());
+                        locator::replication_strategy_params params{new_rf_per_dc_map, old_tablets.tablet_count()};
+                        auto new_strategy = locator::abstract_replication_strategy::create_replication_strategy("NetworkTopologyStrategy", params);
+                        new_tablet_map = co_await new_strategy->maybe_as_tablet_aware()->reallocate_tablets(table, tmptr, old_tablets);
+                    } catch (const std::exception& e) {
+                        error = e.what();
+                        rtlogger.error("Couldn't process global_topology_request::keyspace_rf_change, error: {},"
+                                       "desired rf per dc map: {}", error, new_rf_per_dc_map);
+                        updates.clear(); // remove all tablets mutations ...
+                        break;           // ... and only create mutations deleting the global req
+                    }
+
+                    replica::tablet_mutation_builder tablet_mutation_builder(guard.write_timestamp(), table->id());
+                    auto tablet_id = new_tablet_map.first_tablet();
+                    while (true) {
+                        auto& tablet_info = new_tablet_map.get_tablet_info(tablet_id);
+                        auto last_token = new_tablet_map.get_last_token(tablet_id);
+                        tablet_mutation_builder.set_new_replicas(last_token, tablet_info.replicas)
+                                               .set_stage(last_token, locator::tablet_transition_stage::allow_write_both_read_old)
+                                               .set_transition(last_token, locator::tablet_transition_kind::rf_change);
+                        if (auto next_tablet = new_tablet_map.next_tablet(tablet_id); next_tablet.has_value()) {
+                            tablet_id = *next_tablet;
+                        } else {
+                            updates.emplace_back(tablet_mutation_builder.build());
```

Comment:
please rebase and use `co_await make_canonical_mutation_gently(tablet_mutation_builder.build())` to prevent a potential stall due to large mutation.

Reply:
Done.